modules/fbs-core/math-parser/src/main/kotlin/de/thm/ii/fbs/mathParser/transformers/RuleBasedTransformer.kt

```diff
@@ -0,0 +1,18 @@
+package de.thm.ii.fbs.mathParser.transformers
+
+import de.thm.ii.fbs.mathParser.ast.*
+import de.thm.ii.fbs.mathParser.transformers.rules.Rule
+
+open class RuleBasedTransformer(private vararg val rules: Rule) : BaseTransformer() {
+    override fun transformOperation(input: Operation): Expr {
+        val transformed = super.transformOperation(input)
+
+        for (rule in rules) {
+            if (rule.matches(transformed)) {
+                return rule.apply(transformed)
+            }
+        }
```

Comment:
Instead of searching for the rule each time, you can store it in a map using the init function.

Reply:
This is not possible as `matches` is not a equality check but a abstract method implemented for every rule, which can not be replaced by a map lookup.